drivers/jc42: Add JEDEC JC42.4 temperature sensors

bergzand opened the pull request:
This patch adds a driver for the jc42 compatible
temperature sensors. All sensors compatible with
jc42.4 standard should work. Examples include the mcp9808, 
ADT7408 and more.

### General Device Driver Checklist
- [x] MUST: the supported feature set and any custom behavior is clearly documented
- [x] MUST: device descriptor is defined: `devab_t`
- [x] MUST: device parameter struct is defined: `devab_params_t`
- [x] MUST: a default parameter configuration file is present, e.g. `RIOT/drivers/devab/include/devab_params.h`
- [x] MUST: all error and status return codes are named, e.g. `DEVAB_OK, DEVAB_NOSPI, DEVAB_OVERFLOW, ...`


### Additional Sensor Driver Checklist
- [x] MUST: mandatory device parameters are configurable through this file, e.g. sampling rate, resolution, sensitivity
- [x] MUST: an init function in the style of `int devab_init(devab_t *dev, const devab_params_t *params);` exists
- [x] MUST: after initialization, the device must be operational
- [x] MUST: all error and return values are named, e.g. `DEVAB_OK, DEVAB_NODEV, ...`
- [x] MUST: all 'read' functions return values in their physical representation, e.g. `centi-degree, Pa, lux, mili-G`
- [x] MUST: all 'read' functions return integer values, prefarably `int16_t`
- [x] SHOULD: if multiple dimensions are read, they SHOULD be combined into a data structure
- [x] SHOULD: the driver implements the SAUL interface
- [ ] SHOULD: the driver exports functions for putting it to sleep and waking the device

bergzand commented:
Fixes #6198 of course

aabadie commented on drivers/jc42/include/jc42_internal.h:
```diff
@@ -0,0 +1,50 @@
+ */
+#define JC42_REG_CAP           	        (0x00)
+#define JC42_REG_CONFIG        	        (0x01)
+#define JC42_REG_TEMP_UPPER             (0x02)
```
misalignment

bergzand replied:
Fixed

aabadie commented on drivers/jc42/include/jc42_params.h:
```diff
@@ -0,0 +1,67 @@
+#endif
+
+#define JC42_PARAMS_DEFAULT        {.i2c = JC42_PARAM_I2C_DEV,  \
+                                      .speed    = JC42_PARAM_SPEED, \
```
misalignment

bergzand replied:
fixed

aabadie commented on sys/auto_init/saul/auto_init_jc42.c:
```diff
@@ -0,0 +1,84 @@
+            return;
+        }
+
+
```
extra line should be removed

bergzand replied:
and fixed

bergzand commented:
Done some quick fixes for the minor issues.

aabadie commented on drivers/jc42/jc42.c:
```diff
@@ -0,0 +1,105 @@
+    uint16_t tmp;
+
+    /* Read temperature */
+    if(jc42_get_register(dev, JC42_REG_TEMP, &tmp) != 0) {
```
missing space avec `if`

aabadie commented on drivers/jc42/jc42.c:
```diff
@@ -0,0 +1,105 @@
+    dev->i2c = params->i2c;
+    dev->addr = params->addr;
+    i2c_acquire(dev->i2c);
+    if(i2c_init_master(dev->i2c, params->speed) != 0) {
```
same here

aabadie commented on drivers/jc42/jc42.c:
```diff
@@ -0,0 +1,105 @@
+    i2c_release(dev->i2c);
+
+    /* Poll the device, fail if unavailable */
+    if(jc42_get_config(dev, &config) != 0) {
```
and here

aabadie commented on sys/auto_init/saul/auto_init_jc42.c:
```diff
@@ -0,0 +1,83 @@
+    for (unsigned i = 0; i < JC42_NUMOF; i++) {
+        const jc42_params_t *p = &jc42_params[i];
+        if (jc42_init(&jc42_devs[i], (jc42_params_t*) p) < 0)
+        {
```
the brace should be on the same line as `if`

bergzand commented:
And fixed a few more style comments

aabadie commented on drivers/jc42/jc42.c:
```diff
@@ -0,0 +1,105 @@
+#include "periph/i2c.h"
+#include "byteorder.h"
+#include "xtimer.h"
+#include "saul.h"
```
This include is not needed I think

aabadie commented on drivers/jc42/jc42.c:
```diff
@@ -0,0 +1,105 @@
+#include "saul.h"
+
+#include "jc42.h"
+#include "jc42_params.h"
```
Same for this one

bergzand commented:
I shuffled and removed a few includes, I think I've got them all in the correct place now

aabadie commented:
@bergzand, sorry for the late reply. I just got back to your PR. The driver implementation is fine now. I trust you that it should work.
There's still one thing missing : a test. Can you write a test applicatio in `tests/jc42` ? That would be really great. You can take inspiration on the other ones there.

bergzand commented:
@aabadie I added a test for the driver based on the si70xx test. It should print the temperature every second.

Furthermore, I fixed a bug and added `FEATURES_REQUIRED += periph_i2c` to the Makefiles.dep. Is this last change correct and required?

aabadie commented:
Murdock failure is unrelated (see #6356). It should be fine after a rebase to latest master.

bergzand commented:
Rebased, lets see if Murdock is happy now
